rm/rm-node/src/main/java/org/ow2/proactive/resourcemanager/utils/RMNodeStarter.java

```diff
@@ -432,6 +435,63 @@ public static void main(String[] args) {
         }
     }
 
+    private static BlockingQueue<KeyPair> keyPairs = new LinkedBlockingQueue<>();
+
+    private static void launchBackgorundKeyPairMaker(final int numberOfWorkers) {
+        final Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                while (!Thread.interrupted()) {
+                    try {
+                        // if we obeserve less than 'numberOfWorkers' elements than we goint to
+                        // add 'numberOfWorkers' more
+                        if (keyPairs.size() < numberOfWorkers) {
+                            RMNodeStarter.generateAndAddKeyPairs(numberOfWorkers);
+                        }
+
+                        // sleep 1 second, ot until notification (from getKeyPair method)
+                        synchronized (keyPairs) {
+                            keyPairs.wait(1000);
+                        }
+                    } catch (InterruptedException | NoSuchAlgorithmException e) {
+                        logger.warn("Something went terribly wrong.", e);
+                        Thread.currentThread().interrupt();
+                    }
+                }
+            }
+        });
+        thread.start();
+    }
+
+    private static void generateAndAddKeyPairs(int numberOfWorkers)
+            throws NoSuchAlgorithmException, InterruptedException {
+        KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
+        keyGen.initialize(1024, new SecureRandom());
+        KeyPair keyPair = keyGen.generateKeyPair();
+        keyPairs.put(keyPair);
+    }
+
+    /**
+     *
+     * @return keyPair from head of the keyPairs queue
+     */
+    public static KeyPair getKeyPair() {
+        // if by change queue is empty then we notify background thread which apparently sleeps
+        // background thread make sure that there are at least 'workers' key pairs.
+        if (keyPairs.isEmpty()) {
```

Comment:
LinkedBlockingQueues are synchronized, I don't understand the need of this notify call.

Reply:
Imagine that queue is empty and there is 900 ms to wait until thread wake up and put keys in the queue. For this case, i notify waiting thread so that we do not wait for nothing.